src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala

```diff
@@ -106,4 +121,287 @@ object RuleUtils {
+      bucketSpec: BucketSpec,
+      plan: LogicalPlan): LogicalPlan = {
+    // Extract top level plan including all required columns for shuffle in its output.
+    // This is located inside this function because of bucketSpec.bucketColumnNames.
+    object ExtractTopLevelPlanForShuffle {
+      type returnType = (LogicalPlan, Seq[Option[Attribute]], Boolean)
+      def unapply(plan: LogicalPlan): Option[returnType] = plan match {
+        case project @ Project(_, Filter(_, LogicalRelation(_: HadoopFsRelation, _, _, _))) =>
+          val indexedAttrs = getIndexedAttrs(project, bucketSpec.bucketColumnNames)
+          Some(project, indexedAttrs, true)
+        case project @ Project(_, LogicalRelation(_: HadoopFsRelation, _, _, _)) =>
+          val indexedAttrs = getIndexedAttrs(project, bucketSpec.bucketColumnNames)
+          Some(project, indexedAttrs, true)
+        case filter @ Filter(_, LogicalRelation(_: HadoopFsRelation, _, _, _)) =>
+          val indexedAttrs = getIndexedAttrs(filter, bucketSpec.bucketColumnNames)
+          Some(filter, indexedAttrs, false)
+        case relation @ LogicalRelation(_: HadoopFsRelation, _, _, _) =>
+          val indexedAttrs = getIndexedAttrs(relation, bucketSpec.bucketColumnNames)
+          Some(relation, indexedAttrs, false)
+      }
+      def getIndexedAttrs(
+          plan: LogicalPlan,
+          indexedColumns: Seq[String]): Seq[Option[Attribute]] = {
+        val attrMap = plan.output.attrs.map(attr => (attr.name, attr)).toMap
+        indexedColumns.map(colName => attrMap.get(colName))
+      }
+    }
+
+    // During merge of the index data with the newly appended data, perform
+    // on-the-fly shuffle of the appended-data with the same partition structure
+    // of index so that we could avoid incurring shuffle of the index data
+    // (which is assumed to be the larger sized dataset).
+    //
+    // To do this optimally, we would have to push-down filters/projects before
+    // shuffling the newly appended data to reduce the amount of data that gets
+    // shuffled. However, if project excludes any of the [[indexedColumns]], Spark
+    // handles the shuffle through [[RoundRobinPartitioning]], which can cause
+    // wrong results.
+    //
+    // Therefore, we do the following:
+    // if Project node excludes any of the [[indexedColumns]]:
+    //    Case 1: Shuffle should be located **before** Project but **after** Filter
+    //                Plan: Project => Shuffle => Filter => Relation
+    // else:
+    //    Case 2: Shuffle node should come **after** Project and/or Filter node
+    //                Plan: Shuffle => Project => Filter => Relation
+    //
+    // Currently, we only perform on-the-fly shuffle when applying JoinIndexRule.
+    // Therefore, it's always guaranteed that the children nodes have all indexed columns
+    // in their output; Case 1 won't be shown in use cases. The implementation is kept
+    // for future use cases.
```

Comment:
I couldn't comment on https://github.com/microsoft/hyperspace/pull/165#discussion_r490674500, but is this now being tested? If it's not covered, then I wouldn't add the case and just throw if this case is hit.

Reply:
This behavior is covered by below ` test("Verify the location of injected shuffle for Hybrid Scan.") {`